lib/block-parser.php

```diff
@@ -0,0 +1,205 @@
+<?php
+
+if (!class_exists('Gutenberg_Block_Parser_State', false)):
+
+class Gutenberg_Block_Parser_State {
+	public $block_stack;
+}
+
+endif;
+
+if (!class_exists('Gutenberg_Block_Parser', false)):
+
+class Gutenberg_Block_Parser {
+	const BLOCK_COMMENT_OPEN  = '(^<!--)';
+	const BLOCK_COMMENT_CLOSE = '(^/?-->)';
+	const BLOCK_NAME          = '(^[[:alpha:]](?:[[:alnum:]]|/[[:alnum:]])*)i';
+	const BLOCK_ATTRIBUTES    = '(^{(?:((?!}[ \t\r\n]+/?-->).)*)})';
+	const WS                  = '(^[ \t\r\n])';
+	const WSS                 = '(^[ \t\r\n]+)';
+
+	const MAX_RUNTIME         = 1; // give up after one second
+
+	public function parse($input) {
+		$tic = microtime( true );
+
+		// trampoline for stack-safe recursion of the actual parser
+		while ( $this->input && ( microtime( true ) - $tic ) < self::MAX_RUNTIME ) {
+			return $this->proceed( $input );
+		}
```

Comment:
I am not sure about the non-deterministic behavior introduced by the time limit here.  It will likely mean, for example, that some posts parse correctly on one page load but not on the next.  This is a bit worrisome to me, and it would be better to make sure we understand the performance characteristics of our parser(s) well.